### DISCUSSION TITLE
PIO synchronous serial implementation - Word length issue with autopush

### DISCUSSION THREAD

**Discussion by hanshuebner** [micropython/micropython#12014] [RP2040 / Pico] [answered]
Hello,

I am trying to implement a serial interface to read from an ADC (MCP33151). I need to trigger the serial read cycle at a specific point in time under control of another state machine, and using a specific PIO program to perform the reading looked like the quickest option to me.  The PIO program waits for the start signal on PIN_ADC_READ and then is supposed to read 14 bits off PIN_ADC_DIN into the input shift register.  It generates the clock, reading falling edge of the clock signal.  Once 14 bits have been read, the value should be auto-pushed.

The start signal is correctly detected and the clock starts ticking.  Instead of running for 14 cycles, it is running for 84 (14*6) cycles.  If I change the threshold and the bit counter initialization, I still get six times the number of pulses that I expect.  I'm sure it is something simple that I'm missing, but I can't see it.  Is anyone able to point me at my mistake?

Here's what the relevant signals look like (red is PIN_ADC_READ, green is PIN_ADC_CLK):

<img width="1136" alt="image" src="https://github.com/micropython/micropython/assets/108751/6c9d1366-3827-43f6-a21f-b2462e1e932a">

Here's the code:

```python
import rp2
from machine import Pin

SM_CLOCK_MHZ   = 1
WORD_WIDTH     = 14

PIN_TEST_READ  = 3 # externally connected to PIN_ADC_READ

PIN_ADC_READ   = 12
PIN_ADC_DIN    = 13
PIN_ADC_CLK    = 14

SM_ADC         = 5

@rp2.asm_pio(sideset_init=rp2.PIO.OUT_LOW,
             in_shiftdir=rp2.PIO.SHIFT_LEFT,
             autopush=True,
             push_thresh=WORD_WIDTH)
def adc_receive():
    wait(1, pin, 0)            # wait for READ signal
    mov(x, 14)                 # X is our bit counter
    label("loop")
    in_(pins, 1).side(1)       # read one bit from PIN_ADC_DIN to ISR
    jmp(x_dec, "loop").side(0) # next bit
    wait(0, pin, 0)            # make sure READ is inactive


def t():
    read = Pin(PIN_TEST_READ, Pin.OUT)
    adc = rp2.StateMachine(SM_ADC, adc_receive, freq=SM_CLOCK_MHZ * 1_000_000,
                           sideset_base=Pin(PIN_ADC_CLK), in_base=Pin(PIN_ADC_READ))
    adc.active(1)
    read.on()                                               # trigger SM
    read.off()
    print("0x%08x" % (adc.get()))

t()
```

Thanks much!
Hans

**Comment by rkompass** [comment]
It seems to me there is a confusion of PIN_ADC_READ and PIN_ADC_DIN.
At least I noted that you read in from `pins` in `in_(pins, 1)`, which refers to `in_base`, but also wait for pin in `wait(0, pin, 0)`, which also refers to `in_base`. But the waiting should be for another pin, that is PIN_ADC_READ, right?
You might try to change the order of PIN_ADC_READ and PIN_ADC_DIN, so that PIN_ADC_READ is one pin number higher and try with
 `wait(0, pin, 1)`.

**Reply by hanshuebner** [reply]
> It seems to me there is a confusion of PIN_ADC_READ and PIN_ADC_DIN.

Thank you for pointing that out @rkompass!  That at least explains why I did not read anything back.  I've fixed this and now I see that the right number of bits (14) is shifted into the FIFO.  I still don't understand why the number of cycles is off, though.

**Reply by rkompass** [reply, marked as answer]
The `mov(x, 14)` instruction should be a `set(x, 14)`.
According to the [docs](https://docs.micropython.org/en/latest/library/rp2.html) `mov()` has no data as source.
The PIO assembler is not a perfect solution. For saving code it is cast into the python syntax. It should have recognized the
 `mov(.., constant number)` as an error but apparently did not.

**Reply by hanshuebner** [reply]
That makes perfect sense, thank you @rkompass!
